File: 97_interleaving_string.py
def isInterleave(s1: str, s2: str, s3: str) -> bool:
    m, n, s = len(s1), len(s2), len(s3)
    if s != m + n:
        return False
    
    seen = {}
    def dp(i: int, j: int) -> bool:
        # base case
        if i == m:
            if s3[i+j-s:] == s2[j:]:
                return True
            else:
                return False
        if j == n:
            if s3[i+j-s:] == s1[i:]:
                return True
            else:
                return False

        if (i, j) in seen:
            return seen[(i,j)]
        
        interleavable = False
        if s3[i+j] == s1[i]:
            interleavable |= dp(i+1, j)
        if s3[i+j] == s2[j]:
            interleavable |= dp(i, j+1)
        seen[(i,j)] = interleavable
        return interleavable
    
    return dp(0, 0)

s1 = ""
s2 = "b"
s3 = "b"
print(isInterleave(s1, s2, s3))
# Output: true

s1 = "aabcc"
s2 = "dbbca"
s3 = "aadbbcbcac"
print(isInterleave(s1, s2, s3))
# Output: true

s1 = "aabcc"
s2 = "dbbca"
s3 = "aadbbbaccc"
print(isInterleave(s1, s2, s3))
# Output: false

s1 = ""
s2 = ""
s3 = ""
print(isInterleave(s1, s2, s3))
# Output: true

File: 97_interleaving_string.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>
#include <functional>

using namespace std;

class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int m = s1.size();
        int n = s2.size();
        int s = s3.size();

        if (s != m + n) {
            return false;
        }

        unordered_map<long long, bool> seen;

        function<bool(int, int)> dp = [&](int i, int j) -> bool {
            // base case
            if (i == m) {
                return s3.substr(s-i-j) == s2.substr(j);
            }
            if (j == n) {
                return s3.substr(s-i-j) == s1.substr(i);
            }

            long long key = static_cast<long long>(i) << 32 | j;  // Combining i and j into a single key
            if (seen.find(key) != seen.end()) {
                return seen[key];
            }

            bool interleavable = false;
            if (s3[i + j] == s1[i]) {
                interleavable |= dp(i + 1, j);
            }
            if (s3[i + j] == s2[j]) {
                interleavable |= dp(i, j + 1);
            }
            seen[key] = interleavable;

            return interleavable;
        };

        return dp(0, 0);
    }
};

int main() {
    // Testing the function
    cout << Solution().isInterleave("abc", "def", "adbcef") << endl; // Expected: true
    cout << Solution().isInterleave("abc", "def", "adbce") << endl;  // Expected: false
    return 0;
}